src/workerd/jsg/jsg.h

```diff
@@ -537,13 +537,192 @@ using HasGetTemplateOverload = decltype(
   }
 // Adds reflection to a resource type. See PropertyReflection<T> for usage.
 
+#define JSG_TS_ROOT() \
+  registry.registerTypeScriptRoot()
+// Use inside a JSG_RESOURCE_TYPE block to declare that this type should be considered a "root" for
+// the purposes of automatically generating TypeScript definitions. All "root" types and their
+// recursively referenced types (e.g. method parameter/return types, property types, inherits, etc)
+// will be included in the generated TypeScript. For example, the type `ServiceWorkerGlobalScope`
+// should be a "root", as should any non-global type that we'd like include (e.g. `ExportedHandler`
+// and capabilities such as `KvNamespace`).
+//
+// The reason we even have to define roots in the first place (as opposed to just generating TypeScript
+// definitions for all isolate types) is that some types should only be included when certain
+// compatibility flags are enabled. For example, we only want to include the `Navigator` and spec-
+// compliant URL implementation types if `global_navigator` and `url_standard` are enabled respectively.
+
+#define JSG_TS_OVERRIDE(...) \
+ do { \
+  static const char OVERRIDE[] = JSG_STRING_LITERAL(__VA_ARGS__); \
+  registry.template registerTypeScriptOverride<OVERRIDE>(); \
+ } while (false)
+// Use inside a JSG_RESOURCE_TYPE block to customise the generated TypeScript definition for this type.
```

Comment:
Can this comment be pulled out into a doc file somewhere? It's quite long, and somewhat hard to read as a comment

Reply:
Adding documentation/detail to the jsg README.md would be great.